cmd/mistryd/metrics_recorder.go

```diff
@@ -0,0 +1,124 @@
+package main
+
+import (
+	"fmt"
+	"time"
+	"io/ioutil"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var (
+	MetricsRecorderInstance *MetricsRecorder
+	MetricsRecorderOnce sync.Once
+)
+
+// Singleton access
+func GetMetricsRecorder() *MetricsRecorder {
+    MetricsRecorderOnce.Do(func() {
+        MetricsRecorderInstance = NewMetricsRecorder()
+    })
+
+    return MetricsRecorderInstance
+}
+
+type MetricsRecorder struct {
+	cfg *Config
+	BuildsOnHost *prometheus.GaugeVec
+	Builds *prometheus.CounterVec
+	CacheUtilization *prometheus.CounterVec
+	IncrementalUtilization *prometheus.CounterVec
+	BuildTimes *prometheus.HistogramVec
+}
+
+func NewMetricsRecorder() *MetricsRecorder {
+	r := new(MetricsRecorder)
+
+	r.BuildsOnHost = promauto.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "mistry_builds_on_host",
+			Help: "The number of builds currently on the host",
+		},
+		[]string{"project"},
+	)
+
+	r.Builds = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "mistry_builds",
+			Help: "The number of builds",
+		},
+		[]string{"project", "state"},
+	)
+
+	r.CacheUtilization = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "mistry_cache_utilization",
+			Help: "Cache utilization",
+		},
+		[]string{"project", "cache"},
+	)
+
+	r.IncrementalUtilization = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "mistry_incremental_utilization",
+			Help: "Incremental utilization",
+		},
+		[]string{"project"},
+	)
+
+	r.BuildTimes = promauto.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name: "mistry_build_times_seconds",
+			Help: "Build times",
+			Buckets: prometheus.LinearBuckets(100, 100, 3),
+		},
+		[]string{"project"},
+	)
+
+	return r
+}
+
+// A goroutine which runs in the background and continuously (every 5 minutes)
+// collects the data and populates the metrics for prometheus.
+func (r *MetricsRecorder) RecordDiskRelatedMetrics(buildPath string, projectsPath string) {
+	for {
+		projects, _ := ioutil.ReadDir(projectsPath)
+
+		for _, project := range projects {
+			r.RecordBuildOnHost(project.Name(), buildPath)
+		}
+
+		time.Sleep(5 * time.Minute)
+	}
+}
+
+// Read the number of builds for the project by counting the folders under
+// its directories.
+func (r *MetricsRecorder) RecordBuildOnHost(project string, buildPath string) {
```

Comment:
The arguments can be simplified to `project, buildPath string`.

Ditto for all other functions in this file.